File: utils/logger.py
"""
Logger utility for recording user actions
"""

import os
from datetime import datetime


class Logger:
    """Handles logging of user actions to a log file"""
    
    def __init__(self, log_dir="outputs"):
        """Initialize logger with log directory"""
        self.log_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), log_dir)
        self.log_file = os.path.join(self.log_dir, "activity.log")
        
        # Create log directory if it doesn't exist
        os.makedirs(self.log_dir, exist_ok=True)
        
        # Create log file if it doesn't exist
        if not os.path.exists(self.log_file):
            with open(self.log_file, 'w') as f:
                f.write("CryptoLabX Activity Log\n")
                f.write("=" * 70 + "\n")
    
    def log(self, action):
        """
        Log an action with timestamp
        
        Args:
            action (str): The action performed by user
        """
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        log_entry = f"[{timestamp}] Action: {action}\n"
        
        try:
            with open(self.log_file, 'a') as f:
                f.write(log_entry)
        except Exception as e:
            print(f"Warning: Could not write to log file: {e}")
    
    def get_log_path(self):
        """Return the path to the log file"""
        return self.log_file

File: utils/logger.cpp
/**
 * Logger implementation
 */

#include "logger.h"
#include <iostream>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <sys/stat.h>
#include <sys/types.h>

Logger::Logger(const std::string& dir) : logDir(dir), logFile(dir + "/activity.log") {
    // Create directory if it doesn't exist
    #ifdef _WIN32
        _mkdir(logDir.c_str());
    #else
        mkdir(logDir.c_str(), 0755);
    #endif

    initializeLogFile();
}

Logger::~Logger() {}

void Logger::initializeLogFile() {
    std::ifstream checkFile(logFile);
    if (!checkFile.good()) {
        // File doesn't exist, create it with header
        std::ofstream outFile(logFile);
        if (outFile.is_open()) {
            outFile << "CryptoLabX Activity Log" << std::endl;
            outFile << std::string(70, '=') << std::endl;
            outFile.close();
        }
    }
    checkFile.close();
}

void Logger::log(const std::string& action) {
    try {
        std::ofstream logFileStream(logFile, std::ios::app);
        if (!logFileStream.is_open()) {
            std::cerr << "Warning: Could not write to log file" << std::endl;
            return;
        }

        // Get current timestamp
        std::time_t now = std::time(nullptr);
        std::tm* localTime = std::localtime(&now);

        // Format: [YYYY-MM-DD HH:MM:SS] Action: <action>
        logFileStream << "[" 
                      << std::setfill('0') << std::setw(4) << (localTime->tm_year + 1900) << "-"
                      << std::setfill('0') << std::setw(2) << (localTime->tm_mon + 1) << "-"
                      << std::setfill('0') << std::setw(2) << localTime->tm_mday << " "
                      << std::setfill('0') << std::setw(2) << localTime->tm_hour << ":"
                      << std::setfill('0') << std::setw(2) << localTime->tm_min << ":"
                      << std::setfill('0') << std::setw(2) << localTime->tm_sec
                      << "] Action: " << action << std::endl;

        logFileStream.close();
    } catch (const std::exception& e) {
        std::cerr << "Warning: Could not write to log file: " << e.what() << std::endl;
    }
}

std::string Logger::getLogPath() const {
    return logFile;
}
